src/me.rs

```diff
@@ -779,85 +875,452 @@ fn get_best_predictor<T: Pixel>(
+  row: [  1,  1,  1,  0,  0, -1, -1, -1]
+);
+
+/// Perform hexagon search and refine afterwards.
+///
+/// In the hexagon search stage, candidate motion vectors are examined for
+/// improvement to the current search location. The search location is moved to
+/// the best candidate (if any). This is repeated until the search location
+/// stops moving.
+///
+/// Refinement uses a square pattern that fits between the hexagon candidates.
+///
+/// The hexagon pattern is defined by [`HEXAGON_PATTERN`] and the refinement
+/// is defined by [`SQUARE_REFINE_PATTERN`].
+///
+/// `current` provides the initial search location and serves as
+/// the output for the final search results.
+fn hexagon_search<T: Pixel>(
+  fi: &FrameInvariants<T>, po: PlaneOffset, org_region: &PlaneRegion<T>,
+  p_ref: &Plane<T>, current: &mut FullpelSearchResult, bit_depth: usize,
+  pmv: [MotionVector; 2], lambda: u32, mvx_min: isize, mvx_max: isize,
+  mvy_min: isize, mvy_max: isize, bsize: BlockSize,
+) {
+  // The first iteration of hexagon search is implemented separate from
+  // subsequent iterations, which overlap with previous iterations.
+
+  // Holds what candidate is used (if any). This is used to determine which
+  // candidates have already been tested in a previous iteration and can be
+  // skipped.
+  let mut best_cand_idx: usize = 0;
+
+  let mut best_cand: FullpelSearchResult = FullpelSearchResult::empty();
+
+  // First iteration of hexagon search. There are six candidates to consider.
+  for i in 0..6 {
+    let cand_mv = current.mv + HEXAGON_PATTERN[i];
+    let rd = get_fullpel_mv_rd(
+      fi, po, org_region, p_ref, bit_depth, pmv, lambda, false, mvx_min,
+      mvx_max, mvy_min, mvy_max, bsize, cand_mv,
+    );
+
+    if rd.cost < best_cand.rd.cost {
+      best_cand_idx = i;
+      best_cand.mv = cand_mv;
+      best_cand.rd = rd;
+    }
+  }
+
+  // Run additional iterations of hexagon search until the search location
+  // doesn't update.
+  while best_cand.rd.cost < current.rd.cost {
+    // Update the search location.
+    *current = best_cand;
+    best_cand = FullpelSearchResult::empty();
+
+    // Save the index/direction that was taken to the new search location.
+    let center_cand_idx = best_cand_idx;
+
+    // Look only at candidates that don't overlap with previous iterations. This
+    // corresponds with the three offsets (2D) with the closest direction to
+    // that traveled by the previous iteration. HEXAGON_PATTERN has clockwise
+    // order, so the last direction -1, +0, and +1 (mod 6) give the indices for
+    // these offsets.
+    for idx_offset_mod6 in 5..=7 {
+      let i = (center_cand_idx + idx_offset_mod6) % 6;
+      let cand_mv = current.mv + HEXAGON_PATTERN[i];
+
+      let rd = get_fullpel_mv_rd(
+        fi, po, org_region, p_ref, bit_depth, pmv, lambda, false, mvx_min,
+        mvx_max, mvy_min, mvy_max, bsize, cand_mv,
+      );
+
+      if rd.cost < best_cand.rd.cost {
+        best_cand_idx = i;
```

Comment:
best_cand_idx is no longer needed in this loop, correct?

Reply:
No, it's used in every iteration of the while loop.